File: src/InsertionSort/features/InsertionSort.java
package InsertionSort.features;

import java.util.Arrays;

public class InsertionSort {

    private InsertionSort(){}

    public static <E extends Comparable<E>> void sort(E[] arr){

        for(int i = 0; i < arr.length; i ++){

            // 将 arr[i] 插入到合适的位置
            E t = arr[i];
            int j;
            for(j = i; j - 1 >= 0 && t.compareTo(arr[j - 1]) < 0; j --){
                arr[j] = arr[j - 1];
            }
            arr[j] = t;
        }
    }

    private static <E extends Comparable<E>> boolean isSorted(E[] arr){

        for(int i = 1; i < arr.length; i ++)
            if(arr[i - 1].compareTo(arr[i]) > 0)
                return false;
        return true;
    }

    public static void main(String[] args){

        int[] dataSize = {10000, 100000};
        for(int n: dataSize){

            System.out.println("Random Array : ");

            Integer[] arr = ArrayGenerator.generateRandomArray(n, n);
            Integer[] arr2 = Arrays.copyOf(arr, arr.length);
            SortingHelper.sortTest("InsertionSort", arr);
            SortingHelper.sortTest("SelectionSort", arr2);

            System.out.println();

            System.out.println("Ordered Array : ");

            arr = ArrayGenerator.generateOrderedArray(n);
            arr2 = Arrays.copyOf(arr, arr.length);
            SortingHelper.sortTest("InsertionSort", arr);
            SortingHelper.sortTest("SelectionSort", arr2);

            System.out.println();
            /**
             * Random Array :
             * InsertionSort , n = 10000 : 0.170421 s
             * SelectionSort , n = 10000 : 0.111151 s
             *
             * Ordered Array :
             * InsertionSort , n = 10000 : 0.000144 s
             * SelectionSort , n = 10000 : 0.106255 s
             *
             * Random Array :
             * InsertionSort , n = 100000 : 10.850796 s
             * SelectionSort , n = 100000 : 9.702210 s
             *
             * Ordered Array :
             * InsertionSort , n = 100000 : 0.000536 s
             * SelectionSort , n = 100000 : 7.439078 s
             */
        }
    }
}

File: src/MergeSortMore/optimized.by.insertionsort/InsertionSort.py
from ArrayGenerator import ArrayGenerator
from SortingHelper import sort_test
import copy


class InsertionSort:
    @staticmethod
    def sort(arr):
        for i in range(len(arr)):
            # print("i", i)
            t = arr[i]
            j = i
            for j in range(1, i + 1)[::-1]:
                # print("j", j)
                if t < arr[j - 1]:
                    arr[j] = arr[j - 1]
                else:
                    break
                if j == 1:
                    j = j-1
            # print("change_j", j)
            # if arr == [2, 2, 3, 4, 6, 5]:
            #     print("debug")
            # print(arr)
            arr[j] = t
            # print("end_arr", arr)

    @staticmethod
    def sort2(arr,l,r): # TODO 改造插入排序
        for i in range(len(arr)):
            # print("i", i)
            t = arr[i]
            j = i
            for j in range(1, i + 1)[::-1]:
                # print("j", j)
                if t < arr[j - 1]:
                    arr[j] = arr[j - 1]
                else:
                    break
                if j == 1:
                    j = j-1
            # print("change_j", j)
            # if arr == [2, 2, 3, 4, 6, 5]:
            #     print("debug")
            # print(arr)
            arr[j] = t
            # print("end_arr", arr)

if __name__ == '__main__':
    test_list = [6, 4, 2, 3, 1, 5]
    InsertionSort.sort(test_list)
    print(test_list)
    data_size = [10000, 100000]

    for n in data_size:
        # print("Random Array : ")
        # arr = ArrayGenerator.generate_random_array(n, n)
        # arr2 = copy.deepcopy(arr)
        # print(arr2[:10])
        # sort_test("InsertionSort", arr)
        # sort_test("SelectionSort", arr2)

        print("Ordered Array : ")
        arr = ArrayGenerator.generate_ordered_array(n)
        arr2 = copy.deepcopy(arr)
        print(arr2[:10])
        sort_test("InsertionSort", arr)
        sort_test("SelectionSort", arr2)

        """
        InsertionSort, size: 10000 : 0.009609222412109375 s
        SelectionSort, size: 10000 : 3.472014904022217 s
        """